receiver/awslambdareceiver/receiver.go

```diff
@@ -4,23 +4,266 @@
+	}
+
+	encodingExtension, err := loadEncodingExtension[plog.Unmarshaler](host, cfg.S3Encoding, "logs")
+	if err != nil {
+		return nil, err
+	}
+	s3Service, err := s3Provider.GetService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load the S3 service: %w", err)
+	}
+	return newS3Handler(s3Service, set.Logger, encodingExtension.UnmarshalLogs, next.ConsumeLogs), nil
+}
+
+func newMetricsHandler(
+	ctx context.Context,
+	cfg *Config,
+	set receiver.Settings,
+	host component.Host,
+	next consumer.Metrics,
+	s3Provider internal.S3Provider,
+) (lambdaEventHandler, error) {
+	encodingExtension, err := loadEncodingExtension[pmetric.Unmarshaler](host, cfg.S3Encoding, "metrics")
+	if err != nil {
+		return nil, err
+	}
+	s3Service, err := s3Provider.GetService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load the S3 service: %w", err)
+	}
+	return newS3Handler(s3Service, set.Logger, encodingExtension.UnmarshalMetrics, next.ConsumeMetrics), nil
+}
+
+// loadEncodingExtension tries to load an available extension for the given encoding.
+func loadEncodingExtension[T any](host component.Host, encoding, signalType string) (T, error) {
+	var zero T
+	var extensionID component.ID
+	err := extensionID.UnmarshalText([]byte(encoding))
+	if err != nil {
+		return zero, fmt.Errorf("failed to unmarshal identifier: %w", err)
+	}
+	encodingExtension, ok := host.GetExtensions()[extensionID]
+	if !ok {
+		return zero, errors.Join(fmt.Errorf("unknown extension: %s", extensionID), errEncoderNotFound)
+	}
+	unmarshaler, ok := encodingExtension.(T)
+	if !ok {
+		return zero, fmt.Errorf("extension %q is not a %s unmarshaler", encoding, signalType)
+	}
+	return unmarshaler, nil
+}
+
+// detectTriggerType is a helper to derive the eventType based on the payload content.
+// Supported trigger types are:
+// - S3Event
+func detectTriggerType(data []byte) (eventType, error) {
+	if bytes.HasPrefix(data, []byte(`{"Records"`)) {
+		return s3Event, nil
+	}
+
+	// fallback for possible manual trigger cases
+	key, err := extractFirstKey(data)
+	if err != nil {
+		return "", err
+	}
+
+	return "", fmt.Errorf("unknown event type with key: %s", key)
+}
+
+// extractFirstKey extracts the first JSON key from byte array without parsing it.
+func extractFirstKey(data []byte) (string, error) {
+	pos := 0
+	n := len(data)
+
+	// skip any spaces
+	for pos < n && data[pos] <= ' ' {
+		pos++
+	}
+	if pos >= n || data[pos] != '{' {
+		return "", errors.New("invalid JSON payload, failed to find the opening bracket")
+	}
+
+	// advance to opening quote
+	pos++
+	for pos < n && data[pos] <= ' ' {
+		pos++
+	}
+	if pos >= n || data[pos] != '"' {
+		return "", errors.New("invalid JSON payload, expected a key but found none")
+	}
+
+	// extract the first key
+	pos++
+	keyStart := pos
+	for pos < n {
+		if data[pos] == '"' {
+			return string(data[keyStart:pos]), nil
+		}
+		pos++
+	}
+
+	return "", errors.New("invalid JSON payload")
```

Comment:
This can be simplified (untested) using Cut/Trim functions:

```suggestion
	// advance to start of JSON object
	data = bytes.TrimLeft(data, " ")
	data, ok := bytes.CutPrefix(data, []byte{'{'})
	if !ok {
		return "", errors.New("invalid JSON payload, failed to find the opening bracket")
	}

	// advance to opening quote
	data = bytes.TrimLeft(data, " ")
	data, ok = bytes.CutPrefix(data, []byte{'"'})
	if !ok {
		return "", errors.New("invalid JSON payload, expected a key but found none")
	}

	// extract the first key
	data, ok = bytes.Cut(data, []byte{'"'})
	if !ok {
		return "", errors.New("invalid JSON payload")
	}
	return string(data), nil
```

Reply:
Alternatively, we could use the json.Decoder.Token API: https://pkg.go.dev/github.com/goccy/go-json#Decoder.Token. The hit to perf _might_ be noticeable for CloudWatch subscription, unlikely for S3.